crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+# Writing type-checking / type-inference tests
+
+Any Markdown file can be a test suite.
+
+In order for it to be run as one, `red_knot_test::run` must be called with its path; see
+`crates/red_knot_python_semantic/tests/mdtest.rs` for an example that treats all Markdown files
+under a certain directory as test suites.
+
+A Markdown test suite can contain any number of tests. A test consists of one or more embedded
+"files", each defined by a triple-backticks fenced code block. The code block must have a tag string
+specifying its language; currently only `py` (Python files) and `pyi` (type stub files) are
+supported.
+
+The simplest possible test suite consists of just a single test, with a single file.
+
+````markdown
+```py
+reveal_type(1)  # revealed: Literal[1]
+```
+````
+
+When running this test, the mdtest framework will write a file with these contents to the default
+file path (`/src/test.py`) in its in-memory file system, run a type check on that file, and then
+match the resulting diagnostics with the assertions in the test. Assertions are in the form of
+Python comments. If all diagnostics and all assertions are matched, the test passes; otherwise, it
+fails.
+
+## Assertions
+
+Two kinds of assertions are supported, `# revealed:` (shown above) and `# error:`.
+
+### Assertion kinds
+
+#### revealed
+
+A `# revealed:` assertion should always be paired with a call to the `reveal_type` utility, which
+reveals (via a diagnostic) the inferred type of its argument (which can be any expression). The text
+after `# revealed:` must match exactly with the displayed form of the revealed type of that
+expression.
+
+The `reveal_type` function can be imported from the `typing` standard library module (or, for older
+Python versions, from the `typing_extensions` pseudo-standard-library module -- it is a third-party
+module, but typeshed, and thus type checkers also, treat it as part of the standard library):
+
+```py
+from typing import reveal_type
+
+reveal_type("foo")  # revealed: Literal["foo"]
+```
+
+For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
+not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
+without importing it, in addition to the diagnostic revealing the type of the expression.
+
+The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
+undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
+without importing it. (Style preference is to usually not import it in tests, unless specifically
+testing something about the behavior of importing it.)
```

Comment:
I think the current text accurately and precisely describes how they are handled, and this information is relevant to authors/readers of tests, because the details of how they are handled can be visible to authors of tests (for example, if you do `rt = return_type` and then `rt(foo) # revealed: ...` on another line, you will see unimported-reveal diagnostic on the `rt = return_type` line, and it won't be matched by the `# revealed: ...` assertion. And we will have tests that depend on these distinctions (the tests of the undefined-reveal diagnostic itself.)

The proposed new text isn't noticeably shorter, and it has the disadvantage of being wrong :) It's simply not true that "this diagnostic is never emitted" in tests using mdtest.

Reply:
Reading this again, I do think the text could be re-ordered a bit to emphasize the most important thing (`# revealed: ` matches a revealed type from `reveal_type` and then follow up with the stuff about undefined-reveal separately; I'll look at an edit along those lines.